### ISSUE TITLE
Beta 23: How to queue/play an array of track uris?

### ISSUE THREAD

**Issue by ozzotto** [spotify/ios-sdk#778]
The beta23 sdk provides the following methods to play/queue a trackUri:

```objective-c
- (void)playSpotifyURI:(NSString *)spotifyUri startingWithIndex:(NSUInteger)index startingWithPosition:(NSTimeInterval)position callback:(SPTErrorableOperationCallback)block;
- (void)queueSpotifyURI:(NSString *)spotifyUri callback:(SPTErrorableOperationCallback)block;
```

I want to load and play an array of spotify track uris (eg. count =  14) and I do the following.
```objective-c
[spotifyURIs enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
    dispatch_group_t loadSpotifyUriGroup = dispatch_group_create();
    dispatch_group_enter(loadSpotifyUriGroup);
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [audioStreamingController queueSpotifyURI:obj callback:^(NSError *error) {
            dispatch_group_leave(loadSpotifyUriGroup);
        }];
    });
    dispatch_group_wait(loadSpotifyUriGroup, DISPATCH_TIME_FOREVER);
}];
```
I am using `dispatch_group` to make the whole process serial.
If I comment out the `dispatch_after` block and all the next queue call is done immediately after the other I get the following error:

> Error Domain=com.spotify.ios-sdk.playback Code=15 "The request rate limit was reached."

Is there or will be any more efficient way to load and play an array of spotify track uris?

**Comment by Voxar** [comment]
Hi

Anonymous playlists is not supported at the moment and the queue system is not a mean towards that end.
We suggest managing your track list manually, listening to track completion events to play or queue the next track.

Pseudo example:
```
Array tracksToPlay = [...];
starting playback: 
  play first track
  queue second track
when the first track finished playing:
  queue third track
etc...
```

**Comment by asklausen** [comment]
@Voxar What you're suggesting isn't really doable, unless the user stays inside the app at all time. This is never the case though as users will start a song and put the device to sleep and back in their pockets. As users do that, we can queue a max. of 2 songs for them (since there's no way of listening to the event of a completed song in the background) after which playback will seemingly stop.

I think we can all agree that a music player only able to queue 2 songs before the user has to take action again, is just not acceptable. Do Spotify have plans of bringing back basic functionality such as this?

I've tried another method of creating a private playlist in my users Spotify library, to which I replaceURIs() and initiate a track using a specific index. However, I found this method to be extremely unstable as well, as Spotify's API and backend can't detect fast enough that the songs within a playlist has changed. This leads to the SPTAudioStreaming musicPlayer basically initiating Song B even though I verify that the track being fed to Spotify is Song A. It also leads to the musicPlayer very often to play tracks after skipNext() that have previously been part of the playlist, but isn't anymore.

In conclusion @Voxar, it's extremely difficult for a developer to make a professional music player using Spotify's API. Huge chunks of development time is used on hacks/fixes for basic functionality of which one would assume just works coming from a company like Spotify.

Are you guys even working on the mobile SDK anymore?

**Comment by benrosen78** [comment]
@asklausen Did you find a solution to queuing an array of songs?

**Comment by asklausen** [comment]
@benrosen78 Ended up queuing an array of tracks using the 'queueSpotifyURI' method. While the method itself is used for a single URI, I repeat the method for the next URI in my array after a minor delay. It's less than ideal, but as I see it, it's the best thing available using Spotify's API.
